aptos-move/framework/aptos-framework/sources/permissioned_signer.move

```diff
@@ -0,0 +1,620 @@
+        map_or(
+            s,
+            perm,
+            |stored_permission| {
+                is_above(stored_permission, threshold)
+            },
+            false,
+        )
+    }
+
+    public fun check_permission_consume<PermKey: copy + drop + store>(
+        s: &signer, threshold: u256, perm: PermKey
+    ): bool acquires PermissionStorage {
+        if (!is_permissioned_signer(s)) {
+            // master signer has all permissions
+            return true
+        };
+        map_or(
+            s,
+            perm,
+            |stored_permission| {
+                 consume_capacity(stored_permission, threshold)
+            },
+            false,
+        )
+    }
+
+    public(friend) fun capacity<PermKey: copy + drop + store>(
+        s: &signer, perm: PermKey
+    ): Option<u256> acquires PermissionStorage {
+        if (!is_permissioned_signer(s)) {
+            return option::some(U256_MAX)
+        };
+        map_or(
+            s,
+            perm,
+            |stored_permission: &mut StoredPermission| {
+                option::some(match (stored_permission) {
+                    StoredPermission::Capacity(capacity) => *capacity,
+                    StoredPermission::Unlimited => U256_MAX,
+                })
+            },
+            option::none(),
+        )
+    }
+
+    public(friend) fun revoke_permission<PermKey: copy + drop + store>(
+        permissioned: &signer, perm: PermKey
+    ) acquires PermissionStorage {
+        if (!is_permissioned_signer(permissioned)) {
+            // Master signer has no permissions associated with it.
+            return
+        };
+        let addr = permission_address(permissioned);
+        if (!exists<PermissionStorage>(addr)) { return };
+        let perm_storage = &mut borrow_global_mut<PermissionStorage>(addr).perms;
+        let key = copyable_any::pack(perm);
+        if (simple_map::contains_key(perm_storage, &key)) {
+            simple_map::remove(
+                &mut borrow_global_mut<PermissionStorage>(addr).perms,
+                &copyable_any::pack(perm)
+            );
+        }
+    }
+
+    // =====================================================================================================
+    // Native Functions
+    ///
+    /// Check whether this is a permissioned signer.
+    public native fun is_permissioned_signer(s: &signer): bool;
+    /// Return the address used for storing permissions. Aborts if not a permissioned signer.
+    native fun permission_address(permissioned: &signer): address;
+    /// Creates a permissioned signer from an existing universal signer. The function aborts if the
+    /// given signer is already a permissioned signer.
+    ///
+    /// The implementation of this function requires to extend the value representation for signers in the VM.
+    /// invariants:
+    ///   signer::address_of(master) == signer::address_of(signer_from_permissioned_handle(create_permissioned_handle(master))),
+    ///
```

Comment:
Interesting. So you need to be able reconstruct the master signer? What delta this creates in the VM is perhaps not that important for an API user. The invariant is important.

Reply:
I don't think we are reconstructing the master signer here? It's just the address of the permissioned signer should be the same as the master one.